### ISSUE TITLE
[Suggestion] Proioritize view distance the least out of all dynamic options.

### ISSUE THREAD

**Issue by HyperSoop** [Wesley1808/ServerCore#43]
Setting the dynamic view distance max to a 5-digit value and being the only player in a low load area allows the view distance to rise indefinitely, but going into a higher load area (with some mobs ticked around) the view distance is, for some reason, decreased (though it's proven that it has little to no perfomance impact).

**Comment by Wesley1808** [comment]
I would like to make this part configurable at some point, if I can find a clean way to do that.

Regardless though, view distance already has the lowest priority to be modified by the mod.
View distance won't lower unless the simulation distance is at its minimum value, and this works the same for other values.

The current "set in stone order" is:
1. Chunk-tick distance (randomticks & mobspawns)
2. Mobcap multiplier
3. Simulation distance
4. View distance

If it is modified, that would most likely mean your simulation distance is already at or below its set minimum value.
I do wonder how view distance is "proven" to have little to no performance impact, though - because especially when moving around, it does have a major impact on memory usage and networking.

**Comment by HyperSoop** [comment]
@Wesley1808 I think the core issue behind this is that every update it's only one value moved by 1 in the direction of the ideal value, rather than setting it to the perfect value instantly.

Also, there is a way to at least balance out the 4 values - make an option to keep them locked to each other proportionally. Like,  the mobcap is 14% of the chunk tick distance, which is 70% of the simulation distance, which is 35% of the view distance.

**Comment by HyperSoop** [comment]
^^^^ @Wesley1808 please any comment on this?

**Comment by Wesley1808** [comment]
Apologies for the late response.

I'm not entirely sure what you mean specifically by locking them to eachother proportionally, its a bit vague.
The main reason why it only does 1 at a time is because if it changes multiple at a time, from my experience it often overshoots the "exact" value we really want. Especially so with (relatively large) fluctuations in tick time, which are pretty common.

Thats also why I'm not sure if its worth 'trying' to immediately set it to the perfect value instantly, since the calculated value will almost never be perfect. 

The other thing is how much of a requirement it really is to move values by more than one at a time. Back when a similar system I made (back then using scarpet) did exactly that, it would overshoot the 'perfect values' way too often, hence why they're locked to eachother like this.

The only use case I can imagine is if some mod has some machine that is so incredibly laggy that once the chunks containing it are loaded, the server needs a __very__ significant amount more time to process each tick. But in that case it would really be more of an issue with that specific mod, and often changing vanilla settings (or here dynamic settings) wouldn't really change that, besides maybe simulation distance but only to completely deactivate that machine.

**Comment by HyperSoop** [comment]
> Apologies for the late response.
> 
> I'm not entirely sure what you mean specifically by locking them to eachother proportionally, its a bit vague. The main reason why it only does 1 at a time is because if it changes multiple at a time, from my experience it often overshoots the "exact" value we really want. Especially so with (relatively large) fluctuations in tick time, which are pretty common.
> 
> Thats also why I'm not sure if its worth 'trying' to immediately set it to the perfect value instantly, since the calculated value will almost never be perfect.
> 
> The other thing is how much of a requirement it really is to move values by more than one at a time. Back when a similar system I made (back then using scarpet) did exactly that, it would overshoot the 'perfect values' way too often, hence why they're locked to eachother like this.
> 
> The only use case I can imagine is if some mod has some machine that is so incredibly laggy that once the chunks containing it are loaded, the server needs a **very** significant amount more time to process each tick. But in that case it would really be more of an issue with that specific mod, and often changing vanilla settings (or here dynamic settings) wouldn't really change that, besides maybe simulation distance but only to completely deactivate that machine.
> 
> If the main concern is just fluctuating playercounts, I've never really had problems using it for that purpose. At least not on a more vanilla-focused server, unless you consistently have over like 10 players joining and leaving every minute (on default config). If you want it to change settings faster there's also an option for that in the config, pretty sure the default is 15 seconds for everything that isn't viewdistance.

Yes, though it's a bit inconvenient that I can't set it below 1 second.
By locking them to each other I meant what I meant - if one changes, all others adjust proporionally. Would be a nice to have as an option.

**Comment by HyperSoop** [comment]
@Wesley1808 please respond here.

**Comment by Wesley1808** [comment]
I might look into it in the future. But for now I'm rather busy with other things and its not very high priority.

**Comment by Wesley1808** [comment]
I've made a few changes to the dynamic setting manager, you should now be able to (in addition to before):
- Easily disable specific settings from being modified automatically
- Modify the order in which the settings change
- Modify the value that settings increment with each update (ex. instead of decreasing by 1, decrease by 3)

I'm still not quite sure if its any good idea to modify 'linked' settings when its own settings change, makes things quite a bit more complex and I doubt it really helps. 

I would also recommend not updating values every second, because the point of there being some time between updates is that we can measure the MSPT after each change, to see how it affects it. Otherwise that data is more or less going to stay the same and the server is gonna keep on lowering or increasing it without knowing if earlier changes already solved the problem. The new configurable incremental values should be a better alternative.

**Comment by HeTaoyz** [comment]
Hello, I want to use this feature to achieve dynamic view distance. But in my actual testing, if the other three values do not change, the view distance will not change either. Could you please tell me how I can control the view distance separately? Thanks you

![image](https://github.com/user-attachments/assets/2fd687c4-2f27-49ca-bc34-19c0f43c1308)
![image](https://github.com/user-attachments/assets/737e78f0-61fb-44f2-875e-98aed6dee6a3)

**Comment by Wesley1808** [comment]
Looks like currently dynamic settings don't work if you only specify a single entry in that list. 
I'll look into that soon. You could probably temporarily work around it by putting the exact same VIEW_DISTANCE entry in the list twice but thats not exactly pretty.
